keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1158 @@
+TrustAnchorSets object to correspond to a signer.
+
+A TrustAnchorSets object is associated with a signer if its object name is the
+signer name, after applying the following transformation:
+
+```go
+var trustAnchorSetsName strings.Builder
+for _,c := range signerName {
+  switch {
+  case c == '-':
+    trustAnchorSetsName.WriteString("--")
+  case c == '/':
+    trustAnchorSetsName.WriteString("-slash-")
+  default:
+    trustAnchorSetsName.WriteRune(c)
+  }
+}
+```
+
+For example, the signer name `example.com/my-signer` becomes the TrustAnchorSets
+object name `example.com-slash-my--signer`.
+
+Note that this transformation is one to one and reversible, with `-` being used
+as an escape character.  To convert backwards from a TrustAnchorSets object name
+to a signer name:
+
+```go
+var signerName strings.Builder
+var processingEscape bool
+var escapeAccumulator strings.Builder
+for _,c := range trustAnchorSetsName {
+  if processingEscape {
+    switch c {
+    case '-':
+      processingEscape = false
+      switch escapeAccumulator.String() {
+      case "":
+        signerName.WriteRune('-')
+      case "slash":
+        signerName.WriteRune('/')
+      }
+      escapeAccumulator.Reset()
+    default:
+      escapeAccumulator.AddRune(c)
+    }
+  } else {
+    switch c {
+    case '-':
+      processingEscape = true
+    default:
+      signerName.WriteRune(c)
+    }
+  }
+}
+```
+
+#### API Object Definition
+
+```go
+// TrustAnchorSets packages up trust anchors.
+//
+// Each TrustAnchorSets object is owned by a particular signer, which
+// updates the trust anchors as the roots of trust for that signer change.
+//
+// A TrustAnchorSets object wraps up multiple trust anchor sets, each of
+// which in turn is a set of one or more trust anchors (individual
+// certificates).  Most use cases will only require a single trust
+// anchor set, conventionally named `default`.  Multiple uniquely-named trust
+// anchor sets will likely only be required in mutual TLS use cases with
+// federation between multiple meshes.
+type TrustAnchorSets struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  // The status of the TrustAnchorSets object.
+  Status TrustAnchorSetsStatus `json:"status"`
+}
+
+type TrustAnchorSetsStatus struct {
+  // The list of trust anchor sets.
+  Sets []TrustAnchorSet
+
+  // The current conditions of the TrustAnchorSets object.
+  //
+  // There is one canonical condition, "Ready", that the signer uses
+  // to indicate to kubelet that the TrustAnchorSets object is up to date.
+  // This is used to keep pods from starting up with empty trust anchor sets
+  // while the signer is initializing.
+  Conditions []metav1.Condition
+}
+
+type TrustAnchorSet struct {
+  // The unique name of this trust anchor set.
+  Name string
+
+  // The individual trust anchors for this set.  Each entry is a single
+  // PEM-wrapped, DER-formatted X.509 certificate.
+  PEMTrustAnchors []string
```

Comment:
PEM is base64-encoded DER, with a tiny framing layer.  The motivation for choosing PEM is that it's a fairly standard format for receiving and processing certificates and keys (especially in the TLS ecosystem).  Administrators working with these objects are more likely to already have the certificate in PEM format.

That said, I don't really feel strongly either way.

Reply:
We can absolutely teach tools to convert to PEM for consumption; if we do pick DER, we should.
One of the drawbacks with PEM is that you don't know if you have one certificate or many; what order they're in. And two documents can differ as binary but be equivalent.

With DER (possibly plus a canonical base64 encoding on top), you can have a single canonical value, and you can assert that you have exactly one certificate. It's easier to reject a cert for not being valid DER because you can verify that (the validation could be out-of-band, even).

Reply:
There is also a thing that if we specify PEM, and we're documenting that, we have to tell people to take a thing that contains base64-encoded data and then base64 encode it. None of the options here is great to explain though.

Reply:
Why would we have to base64 encode it again if we choose PEM? PEM, including it's headers, are yaml/json safe, and I include PEM files in configmaps on a regular basis.

Reply:
OK, based on some discussion in SIG Auth, I changed this field to be a single string that is expected to contain a PEM bundle (multiple certificates, order-independent).

Reply:
> order-independent

Do we specify that these are expected to be _roots_ or might some of those be intermediates?

Reply:
Using the word "roots" is tricky --- these are expected to be roots *for establishing whether or not a certificate is trusted*.  Nothing prevents the certs you put in here from actually chaining up to another certificate.

Certainly, in most usages, I would expect them to be "true" roots.  But they don't have to be.

Reply:
I'm going to resolve this chain, since we've settled the question of ordering of certificates in the file visible to pods (kubelet will directly write the PEM bundle from the ClusterTrustBundle to the file